addons/project/models/project_task.py

```diff
@@ -716,40 +716,72 @@ def copy_data(self, default=None):
         if not_project_user:
             vals_list = [{k: v for k, v in vals.items() if k in self.SELF_READABLE_FIELDS} for vals in vals_list]
         for task, vals in zip(self, vals_list):
-            has_default_name = bool(default.get('name', ''))
-            if not has_default_name:
-                vals['name'] = _("%s (copy)", task.name)
-                vals['child_ids'] = [child.copy({'name': _("%s (copy)", child.name)}).id for child in task.child_ids]
-            else:
-                vals['child_ids'] = [child.copy({'name': child.name}).id for child in task.child_ids]
+
+            vals['stage_id'] = task.stage_id.id
+            vals['name'] = task.name if self.env.context.get('copy_project') else _("%s (copy)", task.name)
             if task.recurrence_id:
                 vals['recurrence_id'] = task.recurrence_id.copy().id
+            if task.allow_milestones:
+                milestone_mapping = self.env.context.get('milestone_mapping', {})
+                vals['milestone_id'] = milestone_mapping.get(vals['milestone_id'], vals['milestone_id'])
```

Comment:
My stupid brain refuses to work when there's 3 times the same thing on the same line :sweat_smile: 
Why don't we do that?
```suggestion
                vals['milestone_id'] = milestone_mapping.get(vals['milestone_id'], False)
```

Reply:
The copy() method is not used only when a project is copied, it also used when tasks are copied alone or in batch from the view list. In this case, we want to keep the default milestone value to correctly copy it. Take the following case : copy task A, with milestone 1. The expected result is task A (copy), with milestone , since the task A (copy) is inside the same project. On the contrary, if we copy the project P, which contains task A, we dont want the milestone 1 on the task A (copy), we want the milestone 1 (copy). The mapping is there to have the correspondence between milestone 1.id and milestone 1(copy).id. If there are no correspondence, it means that the milestone were not copied from the project P. In this case, we dont want to set any milestone on task A (copy). And the tasks copied from the project.copy() have the guarantee that vals['milestone_id'] is false, since we set it in the 'default' arg of copy_data.

Reply:
ps sorry for the long text :angel:

Reply:
Thanks for the explanation, I get it now :+1: